src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -0,0 +1,201 @@
+  /// broadcast its internal status to the whole clsuter. When it's null, it means there
+  /// is no reporter in this node for this component. This is the place there messages are
+  /// generated.
+  /// \param receiver The receiver is defined to be the module which eventually
+  /// will have the view of of the cluster for this component. It's the place where
+  /// received messages are consumed.
+  ///
+  /// \return true if set successfully.
+  bool SetComponents(RayComponentId cid,
+                     const ReporterInterface *reporter,
+                     ReceiverInterface *receiver);
+
+  /// Get the snapshot of a component for a newer version.
+  ///
+  /// \param cid The component id to take the snapshot.
+  ///
+  /// \return If a snapshot is taken, return the message, otherwise std::nullopt.
+  std::optional<RaySyncMessage> GetSnapshot(RayComponentId cid);
+
+  /// Consume a message. Receiver will consume this message if it doesn't have
+  /// this message.
+  ///
+  /// \param message The message received.
+  ///
+  /// \return true if this node doesn't have message with newer version.
+  bool ConsumeMessage(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Return the cluster view of this local node.
+  const absl::flat_hash_map<std::string, Array<std::shared_ptr<const RaySyncMessage>>>
+      &GetClusterView() const {
+    return cluster_view_;
+  }
+
+ private:
+  /// For local nodes
+  Array<const ReporterInterface *> reporters_ = {nullptr};
+  Array<ReceiverInterface *> receivers_ = {nullptr};
+
+  /// This field records the version of the snapshot that has been taken.
+  Array<int64_t> snapshots_taken_;
+  /// Keep track of the latest messages received.
+  /// Use shared pointer for easier liveness management since these messages might be
+  /// sending via rpc.
+  absl::flat_hash_map<std::string, Array<std::shared_ptr<const RaySyncMessage>>>
+      cluster_view_;
+};
+
+class NodeSyncConnection {
+ public:
+  NodeSyncConnection(RaySyncer &instance,
+                     instrumented_io_context &io_context,
+                     std::string node_id);
+
+  /// Push a message to the sending queue to be sent later. Some message
+  /// might be dropped if the module think the target node has already got the
+  /// information. Usually it'll happen when the message has the same node id
+  /// as the target or the message is sent from this node.
+  ///
+  /// \param message The message to be sent.
+  ///
+  /// \return true if push to queue successfully.
+  bool PushToSendingQueue(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Send the message queued.
+  virtual void DoSend() = 0;
+
+  virtual ~NodeSyncConnection() {}
+
+  /// Return the node id of this sync context.
+  const std::string &GetNodeId() const { return node_id_; }
+
+  /// Handle the udpates sent from this node.
+  ///
+  /// \param messages The message received.
+  void ReceiveUpdate(RaySyncMessages messages);
```

Comment:
I'm a bit confused: is this handling the updates received or sent from this node?

Reply:
It's about Handling the message received. Sent from this node will go with DoSend code path.

Reply:
I now know why I had the confusion: when you say `this node` you mean the remote node of the connection but I interpreted it as the local node.

Given that I was confused as the code reader and the connection has two ends (two nodes), I think it's better to name it `remote_node_id` or something similar instead of just `node_id`